embed/ephy-notification-manager.c

```diff
@@ -0,0 +1,112 @@
+/* -*- Mode: C; tab-width: 2; indent-tabs-mode: nil; c-basic-offset: 2 -*- */
+/*
+ *  Copyright © 2016 Gabriel Ivascu <ivascu.gabriel59@gmail.com>
+ *
+ *  This program is free software; you can redistribute it and/or modify
+ *  it under the terms of the GNU General Public License as published by
+ *  the Free Software Foundation; either version 2, or (at your option)
+ *  any later version.
+ *
+ *  This program is distributed in the hope that it will be useful,
+ *  but WITHOUT ANY WARRANTY; without even the implied warranty of
+ *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ *  GNU General Public License for more details.
+ *
+ *  You should have received a copy of the GNU General Public License
+ *  along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+
+#include "config.h"
+#include "ephy-notification-manager.h"
+
+struct _EphyNotificationManager {
+  GdNotification  parent_instance;
+
+  GtkWidget      *grid;
+};
+
+struct _EphyNotificationManagerClass {
+  GdNotificationClass parent_class;
+};
+
+G_DEFINE_TYPE (EphyNotificationManager, ephy_notification_manager, GD_TYPE_NOTIFICATION);
+
+static GObject *
+ephy_notification_manager_constructor (GType                  type,
+                                       guint                  n_construct_params,
+                                       GObjectConstructParam *construct_params)
+{
+  static GObject *self = NULL;
+
+  if (self == NULL) {
+    self = G_OBJECT_CLASS (ephy_notification_manager_parent_class)->constructor (type,
+                                                                                 n_construct_params,
+                                                                                 construct_params);
+    g_object_add_weak_pointer (self, (gpointer) &self);
+  }
+
+  return g_object_ref_sink (self);
+}
+
+static void
+ephy_notification_manager_dispose (GObject *object)
+{
+  EphyNotificationManager *self = EPHY_NOTIFICATION_MANAGER (object);
+
+  if (ephy_notification_manager_get_children_num (self) > 0)
+    g_list_free (gtk_container_get_children (GTK_CONTAINER (self->grid)));
```

Comment:
Hmmm, this looks complicated and probably wrong. Any chance you can explain this? Since the `GtkGrid` is a child widget of the `EphyNotificationManager`, it and its children should be freed automatically with `EphyNotificationManager`. I'm uncomfortable with freeing the `GtkContainer`'s children out from under it.